lock-api/src/main/java/com/palantir/lock/client/timestampleases/TimestampLeaseAcquirerImpl.java

```diff
@@ -16,28 +16,203 @@
+
+        NamespaceTimestampLeaseRequest request = NamespaceTimestampLeaseRequest.of(List.of(requests));
+        List<TimestampLeaseResponses> response =
+                delegate.acquireTimestampLeases(request).getAlias();
+        if (response.size() != 1) {
+            response.forEach(this::asyncUnlock);
+            throw new SafeRuntimeException(
+                    "Response should include exactly one result",
+                    SafeArg.of("request", request),
+                    SafeArg.of("response", response));
+        }
+        return RequestAndResponse.of(requests, response.get(0));
+    }
+
+    private boolean wasRequestFullyFulfilled(RequestAndResponse requestAndResponse) {
+        Map<TimestampLeaseName, Integer> requests =
+                requestAndResponse.requests().getNumFreshTimestamps();
+        Map<TimestampLeaseName, TimestampLeaseResponse> responses =
+                requestAndResponse.response().getTimestampLeaseResponses();
+
+        if (!requests.keySet().equals(responses.keySet())) {
+            asyncUnlock(requestAndResponse.response());
+            throw new SafeRuntimeException(
+                    "Response lease timestamps need to match request timestamp names exactly",
+                    SafeArg.of("requests", requestAndResponse.requests()),
+                    SafeArg.of("response", requestAndResponse.response()));
+        }
+
+        boolean wasFullyFulfilled = requests.keySet().stream().allMatch(timestampName -> {
+            int requestedTimestamps = requests.get(timestampName);
+            long returnedTimestamps =
+                    responses.get(timestampName).getFreshTimestamps().getCount();
+            return returnedTimestamps >= requestedTimestamps;
+        });
+
+        if (!wasFullyFulfilled) {
+            asyncUnlock(requestAndResponse.response());
+            log.info(
+                    "Timestamp lease request was not fully fulfilled. This should happen infrequently.",
+                    SafeArg.of("requests", requests),
+                    SafeArg.of("responses", responses));
+        }
+
+        return wasFullyFulfilled;
+    }
+
+    private void asyncUnlock(TimestampLeaseResponses responses) {
+        unlocker.unlock(responses.getLeaseGuarantee().getIdentifier());
+    }
+
+    private static Map<TimestampLeaseName, TimestampLeaseResult> createTimestampLeaseResult(
+            Map<TimestampLeaseName, Integer> requestedTimestamps,
+            Map<TimestampLeaseName, TimestampLeaseResponse> responses) {
+        return KeyedStream.stream(responses)
+                .<TimestampLeaseResult>map((timestampName, response) -> {
+                    int requestedForName = requestedTimestamps.get(timestampName);
+                    LongSupplier freshTimestamps = new LimitingLongSupplier(
+                            new ConjureTimestampRangeTimestampSupplier(response.getFreshTimestamps()),
+                            requestedForName);
```

Comment:
Question: should I add a test covering this? I was thinking of a parametrized test with 0, 1, 2 retries where I assert that the timestamps are unique and increasing and that trying to fetch more doesn't work. Most of this is already tested though, so in practice, this would be testing whether the right params are passed (namely `requestedForName`) - which is why I didn't add the test and put up the question.

Reply:
Not quite sure if I understand the concern. Maybe we'll have some sort of integration test later once you've wired up everything?

Reply:
I wouldn't test the retry logic, that's always quite hard to do

Reply:
Oh you managed to write it, excellent